File: notification-service/src/main/java/com/bank/notificationservice/security/config/RequestIdFilter.java
package com.bank.notificationservice.security.config;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.slf4j.MDC;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import java.io.IOException;
import java.util.Optional;
import java.util.UUID;

@Component
public class RequestIdFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
            throws ServletException, IOException {
        String id = Optional.of(request.getHeader("requestId"))
                .orElse(UUID.randomUUID().toString());

        MDC.put("requestId", id);

        try {
            filterChain.doFilter(request, response);
        } finally {
            MDC.remove("requestId");
        }
    }
}

File: core-query/src/main/kotlin/ru/patterns/corequery/security/config/RequestIdFilter.kt
package ru.patterns.corequery.security.config

import org.slf4j.MDC
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import java.util.UUID

class RequestIdFilter : WebFilter {
    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val headerValue = exchange.request.headers.getFirst("requestId")

        MDC.put("requestId", headerValue ?: UUID.randomUUID().toString())

        return chain.filter(exchange)
    }
}